subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/cache/KotlinDslWorkspaceProvider.kt

```diff
@@ -31,11 +34,19 @@ class KotlinDslWorkspaceProvider(
     fileAccessTimeJournal: FileAccessTimeJournal,
     inMemoryCacheDecoratorFactory: InMemoryCacheDecoratorFactory,
     stringInterner: StringInterner
-) : DefaultImmutableWorkspaceProvider(
-    "kotlin-dsl",
-    cacheScopeMapping.getBaseDirectory(null, "kotlin-dsl", VersionStrategy.CachePerVersion),
-    cacheRepository,
-    fileAccessTimeJournal,
-    inMemoryCacheDecoratorFactory,
-    stringInterner
-)
+) : WorkspaceProvider, Closeable {
+    private
+    val delegate = DefaultImmutableWorkspaceProvider.withBuiltInHistory(
+        cacheRepository
+            .cache(cacheScopeMapping.getBaseDirectory(null, "kotlin-dsl", VersionStrategy.CachePerVersion))
+            .withDisplayName("kotlin-dsl"),
+        fileAccessTimeJournal,
+        inMemoryCacheDecoratorFactory,
+        stringInterner
+    )
+
+    override fun <T : Any> withWorkspace(path: String, action: WorkspaceProvider.WorkspaceAction<T>): T =
+        delegate.withWorkspace("$accessorsWorkspacePrefix/$path", action)
```

Comment:
Isn't the cleanup depth wrong here, given that we introduce a directory in between?